jplag.frontend.emf-metamodel/src/main/java/de/jplag/emf/util/MetamodelTreeView.java

```diff
@@ -34,23 +37,29 @@ public MetamodelTreeView(String filePath) {
     }
 
     /**
-     * Adds a token to the view, thus adding the index information to the token.
+     * Adds a token to the view, thus adding the index information to the token. Returns a new token enriched with the index
+     * metadata.
      * @param token is the token to add.
      * @param treeDepth is the current containment tree depth, required for the indentation.
      */
-    public void addToken(MetamodelToken token, int treeDepth, String prefix) {
-        token.getEObject().ifPresent(it -> {
+    public MetamodelToken convertToMetadataEnrichedTokenAndAdd(MetamodelToken token, int treeDepth, String prefix) {
```

Comment:
Probably the best solution for now, but not really a nice solution, as this method deals with two very different concerns, which makes it a but confusing for outsiders.

Reply:
True, but this "two concerns" behaviour was already present before the refactoring. In the old state, the token was modified directly as a side effect instead of creating a new instance. So I would argue that the changes don't make things worse than they already were 😅 .